### ISSUE TITLE
Add `module` block to SDL

### ISSUE THREAD

**Issue by vpetrovykh** [edgedb/edgedb#907]
The current version of SDL and migrations does not allow expressing the migration of interdependent modules because a single SDL blob is only supposed to reference one module. The solution to this is to force SDL to express the entire user-schema in the context of migration. This means that a single SDL block must denote unambiguously which modules the various entities belong to. There are two mechanisms proposed for that:
1) Fully-qualified names for all entities (except ones belonging to `std`, since that's always built-in in all contexts).
2) A special `module <name> { ... }` block. All inner SDL inside this block is understood to have the default module `<name>`. It is possible to have multiple `module` blocks with the same name in a single SDL document.

The total list of modules in the schema is inferred from `module` block names and from fully-qualified names.

Example of SDL using module blocks:
```
module default {
   # the default module here is "default"
   type Foo {
      # the module "std" is also part of the default name 
      # resolution as usual
      property name -> str;
      link bar -> other_mod::Bar;
   }
}

module other_mod {
   # the default module here is "other_mod"
   type Bar {
      link bar -> Bar;
      link foo -> default::Foo;
   }
}
``` 
And the same schema expressed using only fully-qualified names SDL:
```
# only the "std" module can be omitted here, all other 
# names have to be fully-qualified, even "defaut"
type default::Foo {
   property name -> str;
   link bar -> other_mod::Bar;
}

type other_mod::Bar {
   link bar -> other_mod::Bar;
   link foo -> default::Foo;
}
```

The following schemas are also equivalent:
```
# Multiple module blocks with the same name
module default {
   type Foo;
}

module other_mod {
   type Bar;
}

module default {
   type Foo2;
}
```

```
# Each module block has a unique name
module default {
   type Foo;
   type Foo2;
}

module other_mod {
   type Bar;
}
```

```
# Mix of fully-qualified names and module blocks
type default::Foo;

module other_mod {
   type Bar;
}

module default {
   type Foo2;
}
```

**Comment by tailhook** [comment]
It's unclear from the text:
```
type User {
}
module other {
  type User { }
}
```
Is the first `User`'s fully quaified name is `User`? Or `::User`? Or something like in postgres: `public::User`?

> All inner SDL inside this block is understood to have the default module <name>

Does it mean "default module" when referring things. Or when defining too? I mean can I do:
```
module first {
  type other_top_level_module::User {
  }
}
```

**Comment by vpetrovykh** [comment]
The options 1 and 2 are mutually exclusive, meaning:
- A fully qualified name (so a name of the form `<module>::<shortname>`) can only be used in a declaration that is *not* nested in any `module` block. It is also a requirement to use fully-qualified name in this case.
- It is illegal to use fully-qualified name in a declaration inside a `module` block, only short names are allowed.

This is legal:
```
type default::User;
module other {
   type User;
}
```
This is not legal:
```
# this name is not fully-qualified
type User {
}
module other {
  type User { }
}
```

**Comment by tailhook** [comment]
Okay, here is my second take on it.

TL;DR: This text explains three things that are mostly required at the same time:
1. A top-level namespace say `app::`
2. Nested namespaces
3. Everything built-in under `std::`

And here is why.

# Top-Level Namespace

## Rationale

Say you have defined a top-level function:
```
CREATE FUNCTION session() RETURNS Session AS $$...$$ 
```

What's a fully qualified name of it? Is it just `session` ? This basically means there can't be any standard or installable module named `session`. If latter exists, it can't be used.

Note: that some kind of fully-qualified-syntax doesn't work here, because `::session` is the same fully qualified name. 
Note 2: If you say that `::session` can't refer to application types, you have a second problem:
```
type User {}
module something {
  type User {
    parent -> ::User;  # <-- how to refer to the outer user.
  }
}
``` 

A bit offtopic story for Rust:
> In Rust 1.0, you could refer by `user::User` to either current crate's module `user` or to the external crate `user`. In Rust 2018, the latter is referred to by `crate::user`. Because that was a big source of confusion and conflicts. (I can find proof links if needed).
> 
> This is essentially what I'm proposing here as `app::` namespace.

## Proposal

Add default namespace for all the things in current application. I propose name `app::` but we can find a better name.

So basically if you define something like this:
```
type User {}
```
It can be referred to as `app::User` or `::app::User` everywhere.

# Nested Namespaces

If we go with `app::` let's take a look at this fragment:
```
type User {}
module admin {
  type User {
    parent -> app::User;
  }
}
```
This essentially describes nested structures. So fully qualified names should rather be `app::User` and `app::admin::User` rather than `app::User` and `admin::User`. This is also helpful to avoid conflicts with an external package named `admin`.

# The Nested `std::` Namespace

Now, this is twofold, it's both: is backed by the idea of nested namespaces and supports it.

Let's say we want to add module for working with complex numbers:
1. We can add it as `std::complex::Number`
2. Or we can add it as `complex::Number`

Note the first would just work. The second is more problematic. If you already have such a module in your app it's not a problem, because all your things are in `app::`. But if you have an external package with such name it means:
1. Either it can break right away
2. Or we can allow disabling stdlib module in some way
3. Or we can allow renaming stdlib module in some way

Note that (1)-(2) doesn't even allow start using new module to then migrate data. And (3) is kinda working but very confusing.

Note 2: unlike in Rust I expect modules to store data in the DB (i.e. Type is essentially a table, right?), so you can't freely rename imports back and forth.

# Final notes

So I think that all three: `app::` namespace, nested namespaces, and nested std are needed.

Some things like the `cal` module can really be considered as external modules installed everywhere, so basically can be top-level, but adding them later can be a problem.

And also if we're speaking about "packages": yes packages should bring only one top-level namespace (I mean package `x_y` can't inject anything to `x::y::` namespace). But there are lots of legitimate uses on namespaces within the package (I can elaborate more, just this is already a long read, and I think justification above is enough).

**Comment by 1st1** [comment]
> A top-level namespace say app::

This is already the case. We do have a top-level default module called `default::`.  The name is different from what you propose, but the intent is the same.

OT: @vpetrovykh Could you please add a "Name Resolution" section to [EdgeQL/Overview](https://edgedb.com/docs/edgeql/overview)? It looks like the name resolution algorithm and the default module particularities aren't properly documented (or discoverable).

BTW, it is possible to change what's the default module [per query](https://edgedb.com/docs/edgeql/statements/with#specifying-a-module) using `WITH MODULE another_default SELECT Abc` or for the [current connection session](https://edgedb.com/docs/edgeql/statements/sess_set_alias#set), using the `SET MODULE another_default` command.

> Nested namespaces

Yes, let's **leave the door open for them**, meaning that there will be no syntax that prohibits us adding them later in Alpha 2. We don't have engineering capacity to implement them before Alpha 2 or even 1.0. They can be a feature in 2.0 or something.

Having some top-level builtin modules is OK, we won't have too many of them.  [Right now](https://edgedb.com/docs/datamodel/modules) it's `std`, `math`, `sys`, `cfg`.  I imagine we'll add `cal` (soon), `net` (maybe before 1.0) and `gis` (at some point of time **as an optional extension**). Even if we add a couple more in the next few years it shouldn't cause too much problem. 

BTW we can brainstorm and add a list of future-reserved module names; that's quite doable and we can add stuff to the list until 1.0 final is released. 

If we realize that we plan to add a big number of modules in 2.0 we can then add nested namespaces and start compartmentalizing that **new stuff**.

Essentially what I'm trying to say here is that rushing support for nested modules before Alpha 2 isn't necessary.  Although making sure we can add them later **is**.  Let's focus our discussion on that.

> CREATE FUNCTION session() RETURNS Session AS $$...$$ 

This will create a `session()` function in the current **default** module.  All names in EdgeDB are fully qualified, always.

> 1. Either it can break right away
> 2. Or we can allow disabling stdlib module in some way
> 3. Or we can allow renaming stdlib module in some way
>
> [..]
> Note 2: unlike in Rust I expect modules to store data in the DB (i.e. Type is essentially a table, right?), so you can't freely rename imports back and forth.

Option 4: we'll make it possible to create a schema migration that renames a module. That should cover those rare cases when a new version of EdgeDB clashes with some local module. We need to add an `ALTER MODULE` to our existing [module DDL commands](https://edgedb.com/docs/edgeql/ddl/modules).

FWIW there are other scenarios when a user would need to update their code/queries/schema for a new DB version — one of them is when a new reserved keyword is added to the query language. You cannot make it so that it's possible for users to blindly upgrade their DB, unfortunately.

**Comment by tailhook** [comment]
> This is already the case. We do have a top-level default module called default::. The name is different from what you propose, but the intent is the same.

I don't challenge the name. But let's look at adding nested namespaces later:
```
type User {}
module web {
  type Session {}
}
```
So basically we have `default::User` and `web::Session`, right?

Then when we add nested namespaces, we can't add them in rust way:
```
module session {
  module oauth {
  }
}
```
Because that means the user can't make `default::subnamespace::something` (well, it's one look at the things, but inconvenient.

If we add them in C++ way:
```
type User {}
module default::files {
  type 
}
module session::oauth {
  type X {
    user -> User;         # <-- Okay now
    file -> files::File;  # <-- can I refer it like this? Or should I use default::?
  }
}
```
Still keeps a lot of questions on consistency.

I also find `module { module {}}` nicer for name resolution (I can elaborate about it).

>  We don't have engineering capacity to implement them before Alpha 2

Sure, I have thoughts about that. But, let's define goal first, and then cut down MVP.

> Option 4: we'll make it possible to create a schema migration that renames a module.

At a glance, that requires altering all the queries in the app too.

**Comment by 1st1** [comment]
> Because that means the user can't make default::subnamespace::something (well, it's one look at the things, but inconvenient.

Good point! @elprans thoughts?

> At a glance, that requires altering all the queries in the app too.

True.

>> We don't have engineering capacity to implement them before Alpha 2
>
> Sure, I have thoughts about that. But, let's define goal first, and then cut down MVP.

Actually, Elvis, how long do you think we need to implement nested namespaces? If it's on the order of 1-2 days we should consider that.

**Comment by elprans** [comment]
> >  Because that means the user can't make default::subnamespace::something (well, it's one look at the things, but inconvenient.

> Good point! @elprans thoughts?

Not sure I understand the problem.  The "default" module is in no way special, it's just a convention to `SET MODULE default` (like `SET search_path = public` in Postgres).

```
type User {}
module default::files {
  type 
}
```
This is not a legal SDL syntax per the proposal above.  You *must* either qualify the name, or put it inside the `module` block.  I see no conflict with possible future addition of nested modules.

**Comment by 1st1** [comment]
> This is not a legal SDL syntax per the proposal above. You must either qualify the name, or put it inside the module block. I see no conflict with possible future addition of nested modules.

Ah, right.

**Comment by tailhook** [comment]
> type User {}
> module default::files {
>  type 
> }
>
> This is not a legal SDL syntax per the proposal above

I've [explained](https://github.com/edgedb/edgedb/issues/907#issuecomment-559895162) that I don't like that syntax, and you have not addressed that issues.

> Actually, Elvis, how long do you think we need to implement nested namespaces?

Okay, if we can't separate question of what we want implement at all from what we can implement now, here is my approach:

1. Just disable `module {}` for now
2. (optionally) Move all std modules to nested `std::`
3. Rename `default::` to `app::` or whatever (with the current semantics `default::` is okay, but if we put all the user's submodules to this namespace it's ugly)

Now I expect the complexity of nested namespaces come from how we're going to store users namespaces, and how name resolution inside the module works. I.e. moving submodules to std should not be hard.

Okay, as I've said we can keep stdlib modules at top namespace, if we really want to.

Few arguments from the internal chat:
> Python has had no problems introducing new modules to the stdlib.

Well, remember why absolute imports appeared? Before them, there were numerous cases where `import something` was in conflict with stdlib of the same name (I can find examples in popular libraries if you wish).

In python, we don't have this problem specifically because all your own modules are in a single toplevel package/namespace and all imports are absolute.

What you're basically proposing is what was in python before absolute imports:
```
type User {}
module math { # conflict
}
```
In python:
```
app/__init__.py
app/math.py  # no conflict, can import either this or stdlib module math 
```

And yes, fixing this means nested namespaces for app, but yes we can keep stdlib modules in top level.

> > Technically, you can still use math::log() because std:: is always imported, so this isn't a problem.
>
> But then how are nested namespaces improving forward compatibility in any way? :)

Here's how shadowing works:
```
module complex {
}
type Geometry {
  point -> complex::Complex;  # <-- app module complex
}
```
If we introduce `complex` stdlib module, it can be referred only as `::std::complex` (my option) or
`::complex` (flat std namespace). And app's module can be referred to as `::app::complex` (but don't need too).

If you don't put apps' modules to a nestted namespace, your `::complex` is application's module.

**Comment by elprans** [comment]
@tailhook

> I've explained that I don't like that syntax, and you have not addressed that issues.

Please re-read the messages in this thread carefully.  Your criticism is against the syntax that is NOT VALID per this proposal.  Again, the only valid form would be the one, where every definition is explicitly, an unambiguously qualified:

```
type default::User {}

# OR

module default {

  type User {}

}
```

Nesting `module {}` blocks and using fully qualified names to define schema objects inside a `module{}` block is NOT supported.  When, and _if_, we decide to allow nesting modules, we can enable this.


> Rename default:: to app:: or whatever (with the current semantics default:: is okay, but if we put all the user's submodules to this namespace it's ugly)

I'm not against changing the conventional name of the default module.

> If you don't put apps' modules to a nestted namespace, your ::complex is application's module.

We cannot *force* users to put everything under "app", can we? What we _can_ do is to put up a set of recommended practices, and that might just as well recommend using a prefix for application modules, e.g "app_foo", which isn't much different from "app::foo".  This is what people have been happily doing with Django and RoR, both of which have large ecosystems of modules.

I'm not convinced it's worth pursuing an elaborate Rust-like module system at this time.  Nesting and scoping rules bring higher complexity and raise the learning curve, whereas the advantages of formally nested module system vs a flat one are unclear.

I think good documentation and clear conventions, which are, importantly, followed by the client tooling, can achieve the same effect.  For example, the `edgedb migration` tool can adopt a convention of concatenating directory names to form module names, so `schema/app/module/submodule.esdl` becomes `app_module_submodule::` or something like that.

**Comment by tailhook** [comment]
> @tailhook
> 
> > I've explained that I don't like that syntax, and you have not addressed that issues.
> 
> Please re-read the messages in this thread carefully. Your criticism is against the syntax that is NOT VALID per this proposal.

That's entirely true. What I'm arguing there is what syntax would be **if** we allow nesting modules. And that is important for the naming scheme.

Also, you're making an argument that the problem could be solved by a future syntax I'm arguing is bad.

> We cannot force users to put everything under "app", can we?

That's my point. We're discussing two options (are there any third one?):
1. We have a `default` module and `module custom {}`, the latter is top-level
2. We have an `app` module and `module custom {}`, which is `app::custom` (effectively forcing every users's module to appear in the `app`)

> This is what people have been happily doing with Django and RoR, both of which have large ecosystems of modules

I'm not sure what do you mean about django? Are you talking about package names?

*(I don't know much about RoR but Ruby module system is definitely one of the last we want to borrow from)*

> For example, the edgedb migration tool can adopt a convention of concatenating directory names to form module names, so schema/app/module/submodule.esdl becomes app_module_submodule:: or something like that.

So in **your structure**:
1. You start with `schema/app.esdl`
2. Then when you need another module you create it in a dir: `schema/app/something.esdl`
3. Then you might have a conflict with `schema/app_something.esdl`

I would argue that people will often fail on step 2 and just create a top-level namespace (it might happen in few months or few years in development, they already know edgedb and might not reread documentation again).

In **my structure**:
1. You start with `schema/default.esdl` (or a better name, that goes to `app::` directly)
2. Then you add another module `schema/something.esdl` that goes to `app::something`

Looks simpler and cleaner to me.

**Comment by elprans** [comment]
> future syntax I'm arguing is bad

We must be imagining different syntaxes.

In this proposal, we have:

```
module app {

  type User {}

}
```

This defines the object type `app::User`.  

Once nested modules are allowed:

```
module default {
  type User {}

  module web {
    type Session {}
  }
}
```

This defines two object types: `app::User` and `app::web::Session`.  Why is this syntax bad?  This is exactly how namespaces work in Rust and C++.

>    Then you might have a conflict with schema/app_something.esdl

Why would you want both `schema/app_something` and `schema/app/something`?  Just pick one.  This is a non-issue.

**Comment by tailhook** [comment]
> > future syntax I'm arguing is bad
> 
> We must be imagining different syntaxes.
> Once nested modules are allowed:
> 
> ```
> module app {
>   type User {}
> 
>   module web {
>     type Session {}
>   }
> }
> ```
> 
> This defines two object types: `app::User` and `app::web::Session`. Why is this syntax bad?
> 

This is good. And I was arguing that this syntax doesn't allow you adding a subnamespace to `default::` namespace (without adding a special-cased-quirky syntax for that specific thing).

> > Then you might have a conflict with schema/app_something.esdl
> 
> Why would you want both `schema/app_something` and `schema/app/something`? Just pick one. This is a non-issue.

Well:
1. We must check and warn if both files present
2. Every other project may pick a different style
3. Even if i have `schema/app.esdl` and `schema/app/another.esdl` it doesn't forbid my colleague to put his code to `schema/something.esdl`

This is not a big issue per se. But this is just annoyance and subpar design comparing to the alternative.

**Comment by elprans** [comment]
> This is good. And I was arguing that this syntax doesn't allow you adding a subnamespace to default:: namespace 

The `default` module isn't special when it comes to schema definition, so:

```
module default {
   module subns {
   }
}
```

**Comment by tailhook** [comment]
> The default module isn't special when it comes to schema definition, so:

So I probably misunderstand something, that means you can't do:
```
type User {}
```
Which becomes `default::User` ?

**Comment by elprans** [comment]
>  that means you can't do:

Per this proposal, you can't.  SDL definitions are external to the configuration of the database session (including the implicit module alias), and must always state the module explicitly.

**Comment by 1st1** [comment]
> So I probably misunderstand something, that means you can't do:

Yes, it will be illegal under new proposal.  In alpha 2:

```
type User {}
```

will throw an error. In alpha 2 you will need to explicitly write

```
module default {
  type User {};
}
```

**There's only one thing where the 'default' module is special:** in queries, specifically w.r.t. name resolution:

```
SELECT User;
```

is semantically equivalent to

```
WITH MODULE default
SELECT User;
```

We'd like to keep this rule in place because it lowers the bar for people to start playing with EdgeQL.  This will be the only special behavior for 'default' module in alpha 2.

Now, I feel that we're wasting a lot of CPU cycles here. With this proposal we can implement nested modules at any time we wish. Rushing the implementation of nested modules right now does not make a lot of sense.

Probably the only thing left to discuss is the naming of the 'default' module. I see three options:

* Keep the current: 'default'
* Adopt 'app' -- Paul's suggestion. I personally don't like it, while it's better than 'default', not everything is an app.
* Adopt 'main'. Makes sense for apps, for services, for quick scripts.

Ideas?

**Comment by tailhook** [comment]
> Yes, it will be illegal under new proposal. 

Well, sorry. I've somehow skipped this :(

> Probably the only thing left to discuss is the naming of the 'default' module

With this approach `default` is as good as `main`. `app` surely doesn't work.

**Comment by vpetrovykh** [comment]
The new SDL syntax has been implemented.
